x/distribution/keeper/validator.go

```diff
@@ -91,27 +91,36 @@ func (k Keeper) GetValidatorAccum(ctx sdk.Context, operatorAddr sdk.ValAddress)
 	return accum, nil
 }
 
-// updateValidatorDistInfoFromPool updates the validator's distribution info
+// takeValidatorFeePoolRewards updates the validator's distribution info
 // from the global fee pool without withdrawing any rewards. This will be called
 // from a onValidatorModified hook.
-func (k Keeper) updateValidatorDistInfoFromPool(ctx sdk.Context, operatorAddr sdk.ValAddress) sdk.Error {
+func (k Keeper) takeValidatorFeePoolRewards(ctx sdk.Context, operatorAddr sdk.ValAddress) sdk.Error {
 	if !k.HasValidatorDistInfo(ctx, operatorAddr) {
 		return types.ErrNoValidatorDistInfo(k.codespace)
 	}
+	accAddr := sdk.AccAddress(operatorAddr.Bytes())
+
+	// withdraw reward for self-delegation
+	if k.HasDelegationDistInfo(ctx, accAddr, operatorAddr) {
+		fp, vi, di, withdraw :=
+			k.withdrawDelegationReward(ctx, accAddr, operatorAddr)
+		k.SetFeePool(ctx, fp)
+		k.SetValidatorDistInfo(ctx, vi)
+		k.SetDelegationDistInfo(ctx, di)
+		k.WithdrawToDelegator(ctx, fp, accAddr, withdraw)
```

Comment:
Can we get rid of this line? Then the attack in #2914 isn't an issue (and it's also more efficient).